### PR TITLE
VCR: Add trusted issuers

### LINKED ISSUE #134: VCR: add trust to credentials
Add configuration options to trust an issuer of VCs. This must be specified per type.

when untrusted, should it return an error and/or must we provide an override flag?
the same for a credential that is not trusted (yet/anymore)

### PR DESCRIPTION
**Pull request by woutslakhorst**
closes #134 

The trustConfig is configurable via a different file. Location is `data_dir/trusted_issuers.yaml`
There's also an API to add trust/remove trust which also saves the file. 
This comes in handy for testing and on-the-fly adding of trust.
It's expected that a production network doesn't need these but only uses the file.

It's expected that this will change a lot when more different types of VCs and different proof types are added. For now it fills a basic need.

### CODE REVIEW

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: go.mod [hunk 1]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/cbroglie/mustache v1.2.0
 	github.com/deepmap/oapi-codegen v1.4.2
 	github.com/dgrijalva/jwt-go v3.2.1-0.20200107013213-dc14462fd587+incompatible // indirect
+	github.com/ghodss/yaml v1.0.0
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply go.mod:9]: I didn't download any additional so it was probably in our go.sum already through other dependencies.

File: vcr/trust.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+	// to prevent duplicates
+	issuerSet := map[string]bool{
+		issuer.String(): true,
+	}
+	for _, i := range tc.issuesPerType[tString] {
+		issuerSet[i] = true
+	}
+	j := 0
+	var issuerList = make([]string, len(issuerSet))
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vcr/trust.go:103]: That happens when Add/Remove are created first.

File: vcr/trust.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	for k := range issuerSet {
+		issuerList[j] = k
+		j++
+	}
+
+	tc.issuesPerType[tString] = issuerList
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vcr/trust.go:109]: made save private (and placed trust in own package). Placed mutex on other 2 methods

File: vcr/trust.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import (
+	"errors"
+	"os"
+	"sync"
+
+	"github.com/ghodss/yaml"
+	"github.com/nuts-foundation/go-did"
+)
+
+var mutex = sync.Mutex{}
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vcr/trust.go:31]: its now in trustconfig

File: go.mod (hunk 1 above)
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply go.mod:9]: changed to gopkg.in/yaml.v2

File: vcr/api/v1/api.go [hunk 5]
```diff
@@ -154,9 +154,61 @@ func (w *Wrapper) Resolve(ctx echo.Context, id string) error {
...
+	case vcr.ErrRevoked:
+		rr.CurrentStatus = revoked
+	}
+
+	return ctx.JSON(http.StatusOK, rr)
+}
+
+func (w *Wrapper) RemoveTrust(ctx echo.Context) error {
+	return changeTrust(ctx, func(cType did.URI, d did.URI) error {
+		return w.R.RemoveTrust(cType, d)
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vcr/api/v1/api.go:180]: I want to use did, but somebody hijacked that as package name

File: docs/_static/vcr/v1.yaml [hunk 6]
```diff
@@ -169,6 +243,19 @@ components:
...
+      type: object
+      description: result of doing a Resolve operation.
+      required:
+        - currentStatus
+        - verifiableCredential
+      properties:
+        currentStatus:
+          type: string
+          enum: ["trusted", "revoked", "untrusted"]
+          description: trusted is the only valid state. Revoked is returned regardless of if it's trusted.
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply docs/_static/vcr/v1.yaml:256]: If a revoked credential is also untrusted, revoked will be returned

File: docs/_static/vcr/v1.yaml [hunk 7]
```diff
@@ -149,11 +152,82 @@ paths:
...
+          content:
+            text/plain:
+              schema:
+                type: string
 components:
   schemas:
     CredentialSubject:
       type: object
       description: Subject of a Verifiable Credential identifying the holder and expressing claims.
+    IssuerCredentialCombination:
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply docs/_static/vcr/v1.yaml:217]: sure

File: vcr/api/v1/api.go [hunk 8]
```diff
@@ -154,9 +154,61 @@ func (w *Wrapper) Resolve(ctx echo.Context, id string) error {
...
+	if err != nil {
+		return ctx.String(http.StatusBadRequest, fmt.Sprintf("failed to parse credential type: %s", err.Error()))
+	}
+
+	if err = f(*cType, *d); err != nil {
 		return err
 	}
 
-	return ctx.JSON(http.StatusOK, vc)
+	return ctx.NoContent(http.StatusAccepted)
```
Comment (reinkrul) [inline vcr/api/v1/api.go:213]: huh? An HTTP response code as body of a HTTP response?

File: vcr/trust/trust.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+func NewConfig(filename string) Config {
+	return Config{
+		filename:       filename,
+		issuersPerType: map[string][]string{},
+		mutex:          sync.Mutex{},
+	}
+}
+
+// Load the trusted issuers per credential type from file
+func (tc Config) Load() error {
```
Comment (reinkrul) [inline vcr/trust/trust.go:48]: now the mutex is part of the struct you pass it by value when calling `Load`, `AddTrust` or `RemoveTrust` which IntelliJ warns about:

`'Load' passes lock by value: type 'Config' contains 'sync.Mutex' which is 'sync.Locker' `

You have to make define it as pointer in the struct or use the pointer receiver on the funcs

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: vcr/api/v1/api.go (hunk 8 above)
Comment (woutslakhorst) [reply vcr/api/v1/api.go:213]: Ehhh, with `NoContent()` you still have to pass the status code as single arg. Does need to be StatusNoContent though.

File: vcr/trust/trust.go (hunk 9 above)
Comment (woutslakhorst) [reply vcr/trust/trust.go:48]: I'll pay closer attention to the tiny picto in the upper right...

Comment (reinkrul) [reply vcr/trust/trust.go]: :-p

**Review by reinkrul — APPROVED** [review]

**Pull request merged**